// @babylonjs/core/Engines/WebGPU/webgpuHardwareTexture.js
import { InternalTextureSource } from "../../Materials/Textures/internalTexture.js";
import { Scalar } from "../../Maths/math.scalar.js";
import * as WebGPUConstants from "./webgpuConstants.js";
/** @internal */
export class WebGPUHardwareTexture {
    constructor(existingTexture = null) {
        this.format = WebGPUConstants.TextureFormat.RGBA8Unorm;
        this.textureUsages = 0;
        this.textureAdditionalUsages = 0;
        this._webgpuTexture = existingTexture;
        this._webgpuMSAATexture = null;
        this.view = null;
        this.viewForWriting = null;
    }
    get underlyingResource() {
        return this._webgpuTexture;
    }
    get msaaTexture() {
        return this._webgpuMSAATexture;
    }
    set msaaTexture(texture) {
        this._webgpuMSAATexture = texture;
    }
    set(hardwareTexture) {
        this._webgpuTexture = hardwareTexture;
    }
    setUsage(textureSource, generateMipMaps, isCube, width, height) {
        generateMipMaps = textureSource === InternalTextureSource.RenderTarget ? false : generateMipMaps;
        this.createView({
            format: this.format,
            dimension: isCube ? WebGPUConstants.TextureViewDimension.Cube : WebGPUConstants.TextureViewDimension.E2d,
            mipLevelCount: generateMipMaps ? Scalar.ILog2(Math.max(width, height)) + 1 : 1,
            baseArrayLayer: 0,
            baseMipLevel: 0,
            arrayLayerCount: isCube ? 6 : 1,
            aspect: WebGPUConstants.TextureAspect.All,
        });
    }
    createView(descriptor, createViewForWriting = false) {
        this.view = this._webgpuTexture.createView(descriptor);
        if (createViewForWriting && descriptor) {
            const saveNumMipMaps = descriptor.mipLevelCount;
            descriptor.mipLevelCount = 1;
            this.viewForWriting = this._webgpuTexture.createView(descriptor);
            descriptor.mipLevelCount = saveNumMipMaps;
        }
    }
    reset() {
        this._webgpuTexture = null;
        this._webgpuMSAATexture = null;
        this.view = null;
        this.viewForWriting = null;
    }
    release() {
        this._webgpuTexture?.destroy();
        this._webgpuMSAATexture?.destroy();
        this._copyInvertYTempTexture?.destroy();
        this.reset();
    }
}
